weld/transforms/size_inference.rs

```diff
@@ -47,11 +48,69 @@ fn func_has_simple_merge(expr: &TypedExpr) -> bool {
 /// Infers the size of an `Appender` in a `For` loop.
 pub fn infer_size(expr: &mut Expr<Type>) {
     expr.transform_up(&mut |ref mut expr| {
+        let mut sym_gen = SymbolGenerator::from_expression(&expr);
         if let For { ref mut iters, ref mut builder, ref mut func } = expr.kind {
-            if iters.len() > 0 && iters.iter().all(|ref iter| iter.is_simple()) && func_has_simple_merge(func) {
-                if let Ident(_) = iters[0].data.kind {
-                    if let Some(newbuilder) = newbuilder_with_size(builder, iters[0].data.as_ref()) {
-                        *builder = Box::new(newbuilder);
+            /* Without this condition, the transform_up calls seems to end up in an infinite recursive loop */
+            if let NewBuilder(None) = builder.kind {
+                if iters.len() > 0 && func_has_simple_merge(func) {
+                    if let Ident(_) = iters[0].data.kind {
+                        /* Need the data_sym var after the length has been determined to update the
+                         * expression. SymbolGenerator seems to borrow expression immutably, so we can't borrow it mutably before.*/
+                        let data_sym = sym_gen.new_symbol("data1");
+                        let (length, data_expr) = if iters.iter().all(|ref iter| iter.start.is_none()) {
+                            /* In this scenario, to determine the length of the appender, we use:
+                             *      length = len(iters[0].data);
+                             * Here, an issue is if iters[0].data is a more complicated expression,
+                             * then we will be re-evaluating it in order to get to its length.
+                             * Thus, we do something like:
+                             *      let data1 = iters[0].data;
+                             *      for(data1, len(data1), ....do appender stuff)
+                             * Note: to extract the let statements, let_expr needs the expression
+                             * in which the let statement is being defined. So first, we need to
+                             * define new_loop -- but to correctly define new loop, we require the
+                             * length -- thus we extract the let statements only at the end after
+                             * new loop has been defined.
+                             */
+                            let data_expr = ident_expr(data_sym.clone(), iters[0].data.ty.clone()).unwrap();
+                            (length_expr(data_expr.clone()).unwrap(), Some(data_expr))
+                        //} else if iters[0].kind == IterKind::RangeIter || iters[1].kind == IterKind::RangeIter {
+                        } else if iters.iter().any(|ref iter| iter.kind == IterKind::RangeIter) {
```

Comment:
@sppalkia I wasn't sure what's the best way to deal with the RangeIter case - so just return None for now since it always works - but this should maybe be changed?